lib/concurrent/configuration.rb

```diff
@@ -1,124 +1,296 @@
 require 'thread'
-require 'concurrent/delay'
+require 'concurrent/lazy_reference'
+require 'concurrent/atomics'
 require 'concurrent/errors'
-require 'concurrent/atomic'
-require 'concurrent/executor/immediate_executor'
-require 'concurrent/executor/thread_pool_executor'
-require 'concurrent/executor/timer_set'
+require 'concurrent/executors'
 require 'concurrent/utility/processor_count'
 
 module Concurrent
   extend Logging
 
-  # A gem-level configuration object.
-  class Configuration
+  # Suppresses all output when used for logging.
+  NULL_LOGGER = lambda { |level, progname, message = nil, &block| }
 
-    # a proc defining how to log messages, its interface has to be:
-    #   lambda { |level, progname, message = nil, &block| _ }
-    attr_accessor :logger
+  # initialize the global executors
+  class << self
+
+    # @!visibility private
+    @@global_logger = Atomic.new(NULL_LOGGER)
+
+    # @!visibility private
+    @@auto_terminate_global_executors = AtomicBoolean.new(true)
+
+    # @!visibility private
+    @@auto_terminate_all_executors = AtomicBoolean.new(true)
+
+    # @!visibility private
+    @@global_fast_executor = LazyReference.new do
+      Concurrent.new_fast_executor(
+        stop_on_exit: @@auto_terminate_global_executors.value)
```

Comment:
The `at_exit` handler checks the auto-termination setting again. The best practice should be to set auto-termination at application start, before the pools warm up. Doing so will prevent the `at_exit` handlers from being created. Should the user disable auto-termination *after* one or more pools warm up, the `at_exit` handler will behave as expected because it checks again. So it's safe either way, but slightly more efficient in the former case.